# Case Study: Intelligent Agents and Rational Decision-Making

## 📚 Learning Objectives

By completing this notebook, you will:
- Analyze real-world systems as intelligent agents
- Map agent design to environment and performance metrics

## 🔗 Where this fits

**Builds on:** Unit 1, lesson 04 "Intelligent Agents and Rationality" — the agent taxonomy, now applied to real deployed systems.

**Used later in:** Course 12 (AIAT 126) — Unit 1, where the same "what is the environment, what is the performance measure" analysis frames a project proposal.

---


## 🎯 The reflex agent that sold tens of millions of units

iRobot launched the **Roomba** in **September 2002**. For its first thirteen years
it had no map of your home and no idea where it had been. It used a
sensor-driven pseudorandom coverage strategy: drive, bump into something, turn a
semi-random amount, drive again — and rely on statistics and time to cover the
floor, passing over the average patch several times from several angles. That is a
**simple reflex agent**, and it is the design you are about to build below.

It was only with the **Roomba 980**, announced on **16 September 2015**, that
iRobot put its visual SLAM (simultaneous localisation and mapping) into a consumer
product: a camera-based system that builds a map as it cleans and tracks where it
is, so it can cover a floor once, systematically, and resume after recharging.

Thirteen years is the interesting number. A reflex agent shipped, sold, and did
the job well enough for over a decade, because the *performance measure* was
"floor is clean", not "floor was cleaned efficiently". The upgrade to a
goal-based, map-building agent happened when customers began to measure the second
thing. **Which agent type you need is decided by the performance measure, not by
which one is more sophisticated** — and that is why PEAS starts with P.

### What goes wrong without PEAS

Skip the specification and you will build a vacuum that maximises something nobody
asked for. If "performance" means dirt collected, an agent can learn to spread
dirt so there is more to collect. If it means area covered, it will cover the same
easy corridor forever. Writing P, E, A and S down before you code forces the
argument about what "good" means to happen while it is still cheap.


# Case Study: Intelligent Agents and Rational Decision-Making

## 📚 Learning Objectives

By completing this notebook, you will:
- Analyze real-world intelligent agent systems
- Understand how rational decision-making is implemented in AI systems
- Evaluate agent performance using PEAS framework
- Identify agent types (reflex, goal-based, utility-based, learning)
- Understand the role of environments in agent design

## 📖 Official Structure Reference

**Unit:** Unit 1: Introduction to AI and Applications  
**Practical Content:** Case studies on intelligent agents and rational decision-making  

## Case Study 1: Autonomous Vacuum Cleaner Agent

Let's analyze a simple autonomous vacuum cleaner as an intelligent agent using the PEAS framework.

## 📥 Inputs & 📤 Outputs

**Inputs:** What we use in this notebook

- Libraries and concepts as introduced in this notebook; see prerequisites and code comments.

**Outputs:** What you'll see when you run the cells

- Printed results, figures, and summaries as shown when you run the cells.

---

In [1]:
# Case study: describing a robot vacuum with the PEAS framework (Performance measure,
# Environment, Actuators, Sensors). PEAS is how you specify ANY agent problem before coding it.
# Case Study 1: Autonomous Vacuum Cleaner Agent
# PEAS Framework Analysis

print("=== PEAS Framework Analysis ===")
print("\nPerformance Measure:")
print("- Cleanliness of floors")
print("- Time to complete cleaning")
print("- Energy efficiency")
print("- Battery life")

print("\nEnvironment:")
print("- Type: Discrete, Partially Observable, Sequential")
print("- Characteristics: Grid-based floor layout, dynamic (people, pets)")
print("- Agents: Single agent (vacuum cleaner)")

print("\nActuators:")
print("- Movement motors (forward, backward, turn)")
print("- Vacuum motor (on/off)")
print("- Brush mechanism")

print("\nSensors:")
print("- Dirt sensors")
print("- Bump sensors (collision detection)")
print("- Battery level sensor")
print("- Camera/LIDAR for mapping (in advanced models)")

=== PEAS Framework Analysis ===

Performance Measure:
- Cleanliness of floors
- Time to complete cleaning
- Energy efficiency
- Battery life

Environment:
- Type: Discrete, Partially Observable, Sequential
- Characteristics: Grid-based floor layout, dynamic (people, pets)
- Agents: Single agent (vacuum cleaner)

Actuators:
- Movement motors (forward, backward, turn)
- Vacuum motor (on/off)
- Brush mechanism

Sensors:
- Dirt sensors
- Bump sensors (collision detection)
- Battery level sensor
- Camera/LIDAR for mapping (in advanced models)


In [2]:
# From PEAS description to working code: a simple REFLEX agent — it maps the current
# percept straight to an action with no memory or planning, the simplest agent design.
# Simple Reflex Vacuum Agent Implementation
class SimpleVacuumAgent:
    """Simple reflex-based vacuum cleaner agent"""
    def __init__(self):
        self.location = (0, 0)
        self.battery = 100
        self.cleaned = set()
    
    def sense(self, environment):
        """Sense the environment"""
        x, y = self.location
        return {
            'dirt': environment.get((x, y), False),
            'battery': self.battery,
            'location': (x, y)
        }
    
    def act(self, perception, environment):
        """Simple reflex: if dirty, clean; else move"""
        if perception['battery'] <= 10:
            return 'return_to_charger'
        elif perception['dirt']:
            environment[self.location] = False   # cleaning removes the dirt
            self.cleaned.add(self.location)
            self.battery -= 1
            return 'clean'
        else:
            x, y = self.location
            self.location = (x + 1, y)           # simple sweep: move one tile right
            self.battery -= 1
            return 'move_forward'

# Example usage
# Simulate six steps in a tiny 1-row world and watch the percept -> action pairs.
agent = SimpleVacuumAgent()
environment = {(0, 0): True, (1, 0): False, (2, 0): True}

print("=== Simple Vacuum Agent Decision-Making ===")
for step in range(6):
    perception = agent.sense(environment)
    action = agent.act(perception, environment)
    print(f"Step {step + 1}: Location {perception['location']}, Dirt: {perception['dirt']}, Action: {action}")

print(f"\nTiles cleaned: {sorted(agent.cleaned)}  |  Battery left: {agent.battery}%")

=== Simple Vacuum Agent Decision-Making ===
Step 1: Location (0, 0), Dirt: True, Action: clean
Step 2: Location (0, 0), Dirt: False, Action: move_forward
Step 3: Location (1, 0), Dirt: False, Action: move_forward
Step 4: Location (2, 0), Dirt: True, Action: clean
Step 5: Location (2, 0), Dirt: False, Action: move_forward
Step 6: Location (3, 0), Dirt: False, Action: move_forward

Tiles cleaned: [(0, 0), (2, 0)]  |  Battery left: 94%


## 💬 Discuss

Read the six-step trace above before answering — especially step 6, where the
agent moves to `(3, 0)`, a tile that does not exist in `environment`.

1. Our performance measure was implicitly "clean the dirty tiles". The agent
   cleaned both and then walked off the edge of its world, still burning battery.
   **Write a better performance measure in one sentence** — then find a way an
   agent could score well on your measure while doing something useless. (This is
   reward hacking, and it is not a rare failure.)
2. The Roomba stayed a reflex agent for thirteen commercially successful years.
   Given the trace above, argue whether the vacuum in *this* notebook should be
   upgraded to a goal-based agent, or whether a better reflex rule would do.
   What evidence from the printout supports your side?
3. A robot vacuum with a camera and a home map is a mapping device inside a family's
   house. For a Saudi household, what should the manufacturer be required to
   disclose, and what should the device be forbidden from sending anywhere?
   Where exactly does the boundary between "useful sensor" and "surveillance" sit?


## The Four Classic Agent Types — Where Our Vacuum Fits

| Agent type | How it decides | Our vacuum? |
|---|---|---|
| **Simple reflex** | Condition → action rules on the current percept only | ✅ this is what we built: *if dirt → clean, else → move* |
| **Goal-based** | Chooses actions that lead toward an explicit goal state | A vacuum that plans a route to cover every tile |
| **Utility-based** | Maximizes a utility score (speed vs battery vs coverage) | A vacuum that weighs finishing fast against saving battery |
| **Learning** | Improves its policy from experience | A vacuum that learns which rooms get dirty fastest |

Notice the limitation the demo exposes: our reflex agent only sweeps right and never plans a route — exactly why goal-based agents were invented. You will meet learning agents properly in the Reinforcement Learning course (AIAT 123).

## ⚠️ Where this breaks

**The demo's own failure is the lesson.** Look at the trace: the agent cleaned
`(0,0)`, moved right, cleaned `(2,0)`, and then kept moving right to `(3,0)` —
outside its environment — with 94% battery still to burn. It did that because a
reflex agent has **no memory and no goal**. It cannot know it is finished, cannot
know it has left the room, and cannot plan a route back.

- **No memory means no coverage guarantee.** With one sweep direction and no map,
  any tile the agent starts to the right of is never cleaned. The real Roomba
  survived this only by relying on randomness plus a great deal of time.
- **No model of the environment means no recovery.** Battery ≤ 10 returns
  `return_to_charger` as a *string*. There is no route home, because the agent has
  never represented where home is.
- **Partial observability is the normal case.** The agent senses dirt only on the
  tile it occupies. It cannot see a spill one metre away, so its "environment" is
  a one-tile window that it mistakes for the world.
- **PEAS specifies, it does not verify.** Writing "Performance measure:
  cleanliness of floors" commits you to nothing measurable. Until you say
  *cleanliness measured how, over what area, within what time and energy budget*,
  two engineers can both claim to have met the spec.
- **The environment description here is a static list.** Real homes are
  **dynamic** — people, pets and new dirt arrive while the agent is running — so
  a plan computed once is stale by the time it executes. Dynamic environments are
  why learning agents (Course 09) exist.

**Cheaper alternative:** before adding SLAM, a map and a planner, check whether a
better *reflex rule* plus more running time meets the performance measure. iRobot
shipped exactly that trade for thirteen years, and the extra sensor only earned
its cost when the measure changed.

### 🔁 The same reasoning shows up here too

- **Thermostats and lighting controllers:** reflex rules that were good enough
  until "comfort" was replaced by "kWh per comfortable hour" as the measure.
- **Warehouse robots:** reflex obstacle avoidance underneath a goal-based route
  planner — the same two layers, with the reflex layer as the safety net.
- **Email spam filtering:** a reflex rule ("contains this phrase → junk") that was
  replaced by a learning agent only once spammers made the environment dynamic.


## 📚 References

1. Wooldridge, M., & Jennings, N. R. (1995). *Intelligent Agents: Theory and Practice*. The Knowledge Engineering Review, 10(2), 115–152.
2. Russell, S., & Norvig, P. (2020). *Artificial Intelligence: A Modern Approach* (4th ed.), Ch. 2: Intelligent Agents. Pearson.
3. Wang, L., Ma, C., Feng, X., et al. (2023). *A Survey on Large Language Model based Autonomous Agents*. Frontiers of Computer Science (2024). <https://arxiv.org/abs/2308.11432>